### ISSUE TITLE
Force/Sync configs to disk

### ISSUE THREAD

**Issue by warjort** [TheElectronWill/night-config#111]
Is there a reason why configs are not force synced to disk?

See https://forums.minecraftforge.net/topic/113434-why-arent-config-files-force-synced-to-disk/
for a discussion.

**Comment by TheElectronWill** [comment]
Hi!

1. Calling sync decreases performance. How much? Well, it depends partly on the other applications, but usually: quite a lot.
2. Calling sync forces the data to be written to the disk now. Would that really improve the issue of corrupted files? If the data was kept in the cache and never written, nothing would have been corrupted. Thus, the computed must have crashed *during* the writing. In that case, I'm not sure that sync() would have helped 🤔
3. A better solution to prevent corruption could be to write the config in a different, temporary file, and then to move it to the real location.

Before implementing this, it would be better to test different crash scenarios and run some benchmarks :)

**Comment by TheElectronWill** [comment]
Maybe #64 would help too, since it would avoid writing the config in an intermediary/invalid state (and less writes => lower probability of corruption :tada:).

**Comment by Zyllian** [comment]
Just had almost all my NightConfig configs corrupted when my computer hard crashed during server shutdown. Luckily the configs were backed up so it wasn't too bad. Could be coincidence but nothing was lost except NightConfig configs.

**Comment by TheElectronWill** [comment]
Hello!

If the machine crashed while the configs were being written, I'm afraid it's "expected". With a slightly different timing, it could have happened with any other files (that were being written)...

There is no perfect solution to prevent file corruption in case of abrupt computer shutdown. Simply killing the program (e.g. by the task manager) requires it to do some careful work, and the program relies on the OS to do its work reliably, therefore you can imagine that a power failure or an OS-level crash is much harder to deal with.

Major tech companies still do backups :) It's great that you had some!

See [my comment above](https://github.com/TheElectronWill/night-config/issues/111#issuecomment-1177637343) about potential mitigations. I think that a good way of hardening the config writing would be to write to another file, then "swap" the files. But! This needs some caution, because we don't want to loose both files in case something bad happens, so the "swapping" needs to be atomic. It should be OK with `java.nio.file.Files.move` (with the `ATOMIC_MOVE` flag). It should use the POSIX system call `rename` which performs an atomic replace (cf. [man page](https://man7.org/linux/man-pages/man2/rename.2.html)).

However, note that the atomicity is guaranteed *by the OS*. If it crashes, then some inconsistencies can appear. Good filesystems like journaled ext4 will probably correct those, hopefully.

TL;DR: if the OS cannot do its job then nothing is 100% guaranteed to work. So, NightConfig will improve, but let's still make backups.

**Comment by TheElectronWill** [comment]
Commit ea0d1cace77ba9306c93a41b232b282d8e7bd8cf adds `WritingMode.REPLACE_ATOMIC` :wink:
